src/lib.rs

```diff
@@ -13,24 +13,103 @@
 // under the License.
 
 extern crate fuse;
+extern crate libc;
 #[macro_use] extern crate log;
+extern crate time;
 
+use std::collections::HashMap;
 use std::ffi::OsStr;
 use std::io;
 use std::path::Path;
+use std::sync::{Arc, Mutex};
+use time::Timespec;
+
+mod nodes;
+
+// TODO(jmmv): Make configurable via a flag and store inside SandboxFS.
+pub const TTL: Timespec = Timespec { sec: 60, nsec: 0 };
 
 /// FUSE file system implementation of sandboxfs.
 struct SandboxFS {
+    /// Mapping of inode numbers to in-memory nodes that tracks all files known by sandboxfs.
+    nodes: Arc<Mutex<HashMap<u64, Arc<nodes::Node>>>>,
 }
 
 impl SandboxFS {
     /// Creates a new `SandboxFS` instance.
     fn new() -> SandboxFS {
-        SandboxFS {}
+        let root = {
+            let inode = fuse::FUSE_ROOT_ID;
+            let now = time::get_time();
+            let uid = unsafe { libc::getuid() } as u32;
+            let gid = unsafe { libc::getgid() } as u32;
+            nodes::Dir::new_empty(inode, inode, now, uid, gid)
+        };
+
+        let mut nodes = HashMap::new();
+        nodes.insert(root.inode(), root);
+        SandboxFS {
+            nodes: Arc::from(Mutex::from(nodes)),
+        }
+    }
+
+    /// Gets a node given its `inode`.
+    ///
+    /// We assume that the inode number is valid and that we have a known node for it; otherwise,
+    /// we crash.  The rationale for this is that this function is always called on inode numbers
+    /// requested by the kernel, and we can trust that the kernel will only ever ask us for inode
+    /// numbers we have previously told it about.
+    fn find_node(&mut self, inode: u64) -> Arc<nodes::Node> {
+        let nodes = self.nodes.lock().unwrap();
+        match nodes.get(&inode) {
+            Some(node) => node.clone(),
+            None => panic!("Kernel requested unknown inode {}", inode),
+        }
     }
 }
 
 impl fuse::Filesystem for SandboxFS {
+    fn getattr(&mut self, _req: &fuse::Request, inode: u64, reply: fuse::ReplyAttr) {
+        let node = self.find_node(inode);
+        match node.getattr() {
+            Ok(attr) => reply.attr(&TTL, &attr),
+            Err(e) => reply.error(e.raw_os_error().unwrap()),
```

Comment:
Very uncomfortable with this unwrap. It constrains the nodes heavily.

Reply:
The `reply.error()` calls expect a libc error code (because that's the only thing the kernel understands in reply to one of these hooks) so there is not much more we can do if we don't get such detail...

Note that, in the Go version, we returned arbitrary errors from helper functions and, in the FUSE layer, used a "conversion" function that either extracted libc error codes from some error types or mapped "known" error types to libc error codes. I think that was a mistake: it's pretty hard to know what's going on regarding errors there because this function does a lot of magic. I wouldn't like to repeat this here.

One alternative I can think about is changing my helper functions to return a `Result<?, i32>` so that we push the responsibility of computing the libc error codes to the helpers.

What do you think? Or what other alternative do you have in mind? Remember: I'm very new to Rust so there may not be a good explanation for what I do :-P

Reply:
My concern is that you don't enforce the existence of a `raw_os_error` in the type system, so there's an ugly `unwrap` that is in the "should not happen" category. Honestly, `io::Error` is a horrible type and should be avoided in general. I suggest having a custom error type that focuses on appropriate libc error codes. You can have specific handling in there for the lack of a `raw_os_error` and avoid making these functions more complex.

Reply:
I see; that's a good idea. I've added a `KernelError` type to represent these though I'm not sure about the names I've used in the type and methods. There is no conversion code from `io::Error` yet because it's not necessary, so it'll come in a later change.